src/plots/cartesian/set_convert.js

```diff
@@ -182,8 +182,9 @@ module.exports = function setConvert(ax) {
             // encounters them, ie all the categories from the
             // first data set, then all the ones from the second
             // that aren't in the first etc.
-            // TODO: sorting options - do the sorting
-            // progressively here as we insert?
+            // it is assumed that this function is being invoked in the
+            // already sorted category order; otherwise there would be
+            // a disconnect between the array and the index returned
```

Comment:
@etpinard I also considered doing a sorting a lot earlier (more upstream), but it would have two issues: 1) the code should possibly reorder only *after* it's established that it's a categorical axis (it can be user specified or determined heuristically); 2) by that point there are lots of places where _allData etc. are persisted on objects in the trace order; 3) I believe it would be wrong to change the trace order just for this CR because of the snail trail example and the general possibility that the user does _not_ want to have plotly change the user input order (which is currently implicitly taken as the trace order).

So my current thought is that we need to do the axis tick sorting _downstream_, nearer the point which is responsible for the d3 data binding order, and we have to accommodate for the possibility that new lines introduce new points that have to be inserted in lines that have already been added previously.